lib/Lib.IntTypes.fst

```diff
@@ -1,127 +1,245 @@
 
-let sec_int_t t = pub_int_t t
+#push-options "--max_fuel 0 --max_ifuel 1 --z3rlimit 50"
 
-let sec_int_v #t u = pub_int_v u
+let pow2_2 _ = assert_norm (pow2 2 = 4)
+let pow2_3 _ = assert_norm (pow2 3 = 8)
+let pow2_4 _ = assert_norm (pow2 4 = 16)
 
-let uintv_extensionality #t #l a b =
-  match t with
-  | U1   -> ()
-  | U8   -> UInt8.v_inj a b
-  | U16  -> UInt16.v_inj a b
-  | U32  -> UInt32.v_inj a b
-  | U64  -> UInt64.v_inj a b
-  | U128 -> UInt128.v_inj a b
+let sec_int_t t = pub_int_t t
 
-(* Declared in .fsti: uint8, uint16, uint32, uint64, uint128 *)
+let sec_int_v #t u = pub_int_v u
 
 let secret #t x = x
 
-let uint #t #l x =
+let mk_int #t #l x =
   match t with
   | U1 -> UInt8.uint_to_t x
   | U8 -> UInt8.uint_to_t x
   | U16 -> UInt16.uint_to_t x
   | U32 -> UInt32.uint_to_t x
   | U64 -> UInt64.uint_to_t x
   | U128 -> UInt128.uint_to_t x
+  | S8 -> Int8.int_to_t x
+  | S16 -> Int16.int_to_t x
+  | S32 -> Int32.int_to_t x
+  | S64 -> Int64.int_to_t x
+  | S128 -> Int128.int_to_t x
+
+val v_extensionality:
+   #t:inttype
+ -> #l:secrecy_level
+ -> a:int_t t l
+ -> b:int_t t l
+ -> Lemma
+  (requires v a == v b)
+  (ensures  a == b)
+let v_extensionality #t #l a b =
+  match t with
+  | U1   -> ()
+  | U8   -> UInt8.v_inj a b
+  | U16  -> UInt16.v_inj a b
+  | U32  -> UInt32.v_inj a b
+  | U64  -> UInt64.v_inj a b
+  | U128 -> UInt128.v_inj a b
+  | S8   -> Int8.v_inj a b
+  | S16  -> Int16.v_inj a b
+  | S32  -> Int32.v_inj a b
+  | S64  -> Int64.v_inj a b
+  | S128 -> Int128.v_inj a b
 
-let uintv_injective #t #l a =
-  uintv_extensionality a (uint (uint_v a))
-
-let u16_us x = x
-let u32_ul x = x
-let u64_uL x = x
-
-let u128 x = FStar.UInt128.uint64_to_uint128 (u64 x)
-
-let size_v_size_lemma s = ()
+let v_injective #t #l a =
+  v_extensionality a (mk_int (v a))
 
-let uint_v_size_lemma s = ()
+let v_mk_int #t #l n = ()
 
 let size_to_uint32 x = x
 
 let size_to_uint64 x = FStar.Int.Cast.uint32_to_uint64 x
 
 let byte_to_uint8 x = x
 
-let cast #t #l t' l' u  =
+let byte_to_int8 x = FStar.Int.Cast.uint8_to_int8 x
+
+#push-options "--z3rlimit 200"
+
+let cast #t #l t' l' u =
+  assert_norm (pow2 8 = 2 * pow2 7); 
+  assert_norm (pow2 16 = 2 * pow2 15); 
+  assert_norm (pow2 128 = 2 * pow2 127); 
+  let open FStar.Int.Cast in
   match t, t' with
   | U1, U1 -> u
   | U1, U8 -> u
-  | U1, U16 -> FStar.Int.Cast.uint8_to_uint16 u
-  | U1, U32 -> FStar.Int.Cast.uint8_to_uint32 u
-  | U1, U64 -> FStar.Int.Cast.uint8_to_uint64 u
-  | U1, U128 -> FStar.UInt128.uint64_to_uint128 (FStar.Int.Cast.uint8_to_uint64 u)
-  | U8, U1 -> FStar.UInt8.rem u 2uy
+  | U1, U16 -> uint8_to_uint16 u
+  | U1, U32 -> uint8_to_uint32 u
+  | U1, U64 -> uint8_to_uint64 u
+  | U1, U128 -> UInt128.uint64_to_uint128 (uint8_to_uint64 u)
+  | U1, S8 -> uint8_to_int8 u
+  | U1, S16 -> uint8_to_int16 u
+  | U1, S32 -> uint8_to_int32 u
+  | U1, S64 -> uint8_to_int64 u
+  | U1, S128 -> pow2_le_compat (bits t' - 1) (bits t); mk_int #S128 #l' (v u)
```

Comment:
Not extractable because it uses mathematical integers. Same for every other case using `v u`